test/python/transpiler/test_clifford_t_passmanager.py

```diff
@@ -0,0 +1,233 @@
+
+        basis_gates = ["cx", "s", "sdg", "h", "t", "tdg"]
+        pm = generate_preset_pass_manager(
+            basis_gates=basis_gates, optimization_level=optimization_level
+        )
+        transpiled = pm.run(qc)
+        self.assertLessEqual(set(transpiled.count_ops()), set(basis_gates))
+
+    @data(0, 1, 2, 3)
+    def test_iqp(self, optimization_level):
+        """Clifford+T transpilation of IQP circuits."""
+        interactions = np.array([[6, 5, 1], [5, 4, 3], [1, 3, 2]])
+        qc = iqp(interactions)
+
+        basis_gates = ["cx", "s", "sdg", "h", "t", "tdg"]
+        pm = generate_preset_pass_manager(
+            basis_gates=basis_gates, optimization_level=optimization_level
+        )
+        transpiled = pm.run(qc)
+        self.assertLessEqual(set(transpiled.count_ops()), set(basis_gates))
+
+        # The transpiled circuit should be fairly efficient in terms of T/Tdg gates:
+        # the circuit contains 3 powers of CZ-gates, each leading to at most 9 T/Tdg gates,
+        # and 3 powers of T-gates, each leading to at most 4 T/Tdg gates.
+        # Importantly, the transpilation should not make this worse.
+        max_t_size = 3 * 9 + 3 * 4
+        transpiled_ops = transpiled.count_ops()
+        t_size = transpiled_ops.get("t", 0) + transpiled_ops.get("tdg", 0)
+        self.assertLessEqual(t_size, max_t_size)
+
+    @data(0, 1, 2, 3)
+    def test_graph_state(self, optimization_level):
+        """Clifford+T transpilation of graph-state circuits."""
+        adjacency_matrix = [
+            [0, 1, 0, 0, 1],
+            [1, 0, 1, 0, 0],
+            [0, 1, 0, 1, 0],
+            [0, 0, 1, 0, 1],
+            [1, 0, 0, 1, 0],
+        ]
+        graph_state_gate = GraphStateGate(adjacency_matrix)
+
+        qc = QuantumCircuit(5)
+        qc.append(graph_state_gate, [0, 1, 2, 3, 4])
+
+        basis_gates = ["cx", "s", "sdg", "h", "t", "tdg"]
+        pm = generate_preset_pass_manager(
+            basis_gates=basis_gates, optimization_level=optimization_level
+        )
+        transpiled = pm.run(qc)
+        transpiled_ops = transpiled.count_ops()
+        self.assertLessEqual(set(transpiled_ops), set(basis_gates))
+
+        # The resulting circuit should not have any T/Tdg-gates.
+        t_size = transpiled_ops.get("t", 0) + transpiled_ops.get("tdg", 0)
+        self.assertEqual(t_size, 0)
+
+    @data(0, 1, 2, 3)
+    def test_ccx(self, optimization_level):
+        """Clifford+T transpilation of the Toffoli circuit."""
+        qc = QuantumCircuit(3)
+        qc.ccx(0, 1, 2)
+
+        basis_gates = ["cx", "h", "t", "tdg"]
+        pm = generate_preset_pass_manager(
+            basis_gates=basis_gates, optimization_level=optimization_level
+        )
+        transpiled = pm.run(qc)
+
+        # Should get the efficient decomposition of the Toffoli gates into Clifford+T.
+        self.assertEqual(transpiled.count_ops(), {"cx": 6, "t": 4, "tdg": 3, "h": 2})
+
```

Comment:
these are very nice examples! perhaps we can consider adding more advanced ones, like c3x, c4x, mcx or certain adders?

Reply:
As I have already mentioned somewhere, producing _efficient_ circuits is something we want to do, but this requires a bit more work that we want to do in a follow-up. First, we need to make sure that all the standard definitions for gates like c3x, c4x only use Clifford+T gates (I have an open PR on that somewhere), and then we need to update HighLevelSynthesis plugins to always choose Clifford+T synthesis algorithms when available.